File: src/Tnzi.AI/Entities/SubAgentType.cs
namespace Tnzi.AI.Entities;

/// <summary>
/// 子 Agent 类型定义实体 — 持久化 SubAgentRegistry 中的自定义类型
/// </summary>
public class SubAgentType : AuditedEntity<Guid>, IMultiTenant
{
    /// <summary>Type name (unique identifier, e.g. code-reviewer)</summary>
    public string Name { get; set; } = string.Empty;

    /// <summary>Type description</summary>
    public string Description { get; set; } = string.Empty;

    /// <summary>Tool groups（JSON 值转换列，与 Agent.Domains/Roles 同模式）</summary>
    public List<string>? ToolGroups { get; set; }

    /// <summary>Excluded tool groups（JSON 值转换列）</summary>
    public List<string>? ExcludedToolGroups { get; set; }

    /// <summary>Maximum turns</summary>
    public int MaxTurns { get; set; } = 50;

    /// <summary>System instructions</summary>
    public string? Instructions { get; set; }

    /// <summary>Default model</summary>
    public string? DefaultModel { get; set; }

    /// <summary>Default approval mode</summary>
    public ToolApprovalMode? DefaultApprovalMode { get; set; }

    /// <summary>Capability tags（JSON 值转换列）</summary>
    public List<string>? CapabilityTags { get; set; }

    /// <summary>Whether this type is enabled</summary>
    public bool IsEnabled { get; set; } = true;

    /// <summary>Tenant ID</summary>
    public Guid? TenantId { get; set; }
}

File: src/Tnzi.UI/packages/core/src/services/ai/sub-agent-type.ts
/**
 * Sub-Agent Type API wrappers - CRUD over persisted custom sub-agent type
 * definitions.
 *
 * Mirrors `Tnzi.AI/Controllers/Admin/DefaultSubAgentTypeAdminController` (route
 * `admin/ai/sub-agent-types`). Persisting a type makes it available to the
 * sub-agent registry (e.g. for the `spawn_agent` tool).
 */

import type { HttpClient } from '../../http/http';

// ---------------------------------------------------------------------------
// Enums (mirror backend Tnzi.AI.Options.ToolApprovalMode)
// ---------------------------------------------------------------------------

/** Default tool-approval mode for a sub-agent type. */
export enum ToolApprovalMode {
  /** Never require approval (default). */
  NeverRequire = 0,
  /** Always require approval. */
  AlwaysRequire = 1,
  /** Decide per configured allow/deny lists. */
  Specific = 2,
}

// ---------------------------------------------------------------------------
// DTOs
// ---------------------------------------------------------------------------

/** A persisted sub-agent type definition. */
export interface SubAgentTypeDto {
  id: string;
  /** Type name (unique identifier). */
  name: string;
  /** Type description. */
  description: string;
  toolGroups?: string[] | null;
  excludedToolGroups?: string[] | null;
  maxTurns: number;
  instructions?: string | null;
  defaultModel?: string | null;
  defaultApprovalMode?: ToolApprovalMode | null;
  capabilityTags?: string[] | null;
  isEnabled: boolean;
  creationTime: string;
  lastModificationTime?: string | null;
}

/** Create/update payload for a sub-agent type (shared shape). */
export interface SubAgentTypeInputDto {
  /** Type name (unique identifier). */
  name: string;
  /** Type description. */
  description: string;
  toolGroups?: string[] | null;
  excludedToolGroups?: string[] | null;
  /** Maximum turns (backend default 50). */
  maxTurns?: number;
  instructions?: string | null;
  defaultModel?: string | null;
  defaultApprovalMode?: ToolApprovalMode | null;
  capabilityTags?: string[] | null;
  /** Whether the type is enabled (backend default true). */
  isEnabled?: boolean;
}

// ---------------------------------------------------------------------------
// Admin: Sub-Agent Type API
// Route: /admin/ai/sub-agent-types
// ---------------------------------------------------------------------------

/** Admin sub-agent-type management API */
export function useAdminSubAgentTypeApi(client: HttpClient) {
  const base = '/admin/ai/sub-agent-types';
  return {
    /** Get all sub-agent type definitions */
    getAll: () =>
      client.get<SubAgentTypeDto[]>(base),

    /** Create a sub-agent type definition */
    create: (data: SubAgentTypeInputDto) =>
      client.post<SubAgentTypeDto>(base, data),

    /** Update a sub-agent type definition */
    update: (id: string, data: SubAgentTypeInputDto) =>
      client.put<SubAgentTypeDto>(`${base}/${id}`, data),

    /** Delete a sub-agent type definition */
    delete: (id: string) =>
      client.delete<void>(`${base}/${id}`),
  };
}
